src/stable_hasher.rs

```diff
@@ -8,48 +8,113 @@ use std::hash::Hasher;
 ///
 /// To that end we always convert integers to little-endian format before
 /// hashing and the architecture dependent `isize` and `usize` types are
 /// extended to 64 bits if needed.
+///
+/// # Example
+///
+/// ```
+/// use rustc_stable_hash::{StableHasher, StableHasherResult};
+/// use std::hash::Hasher;
+///
+/// struct Hash128((u64, u64));
+/// impl StableHasherResult for Hash128 {
+///     fn finish(hasher: StableHasher) -> Hash128 {
+///         Hash128(hasher.finalize())
+///     }
+/// }
+///
+/// let mut hasher = StableHasher::new();
+/// hasher.write_usize(0xFA);
+///
+/// let hash: Hash128 = hasher.finish();
+/// ```
+#[must_use]
 pub struct StableHasher {
     state: SipHasher128,
 }
 
-impl fmt::Debug for StableHasher {
-    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
-        write!(f, "{:?}", self.state)
-    }
+/// Trait for retrieving the result of the stable hashing operation.
+///
+/// # Example
+///
+/// ```
+/// use rustc_stable_hash::{StableHasher, StableHasherResult};
+///
+/// struct Hash128((u64, u64));
+///
+/// impl StableHasherResult for Hash128 {
+///     fn finish(hasher: StableHasher) -> Hash128 {
+///         Hash128(hasher.finalize())
+///     }
+/// }
+/// ```
+pub trait StableHasherResult: Sized {
+    /// Retriving the [`finalize`]d state of the [`StableHasher`] and construct
+    /// an [`Self`] containing the hash.
+    ///
+    /// [`finalize`]: StableHasher::finalize
+    fn finish(hasher: StableHasher) -> Self;
 }
 
 impl StableHasher {
+    /// Creates a new [`StableHasher`].
+    ///
+    /// To be used with the [`Hasher`] implementation and [`StableHasher::finish`].
     #[inline]
+    #[must_use]
     pub fn new() -> Self {
         StableHasher {
             state: SipHasher128::new_with_keys(0, 0),
         }
     }
 
+    /// Returns the typed-hash value for the values written.
+    ///
+    /// The resulting typed-hash value is constructed from an
+    /// [`StableHasherResult`] implemenation.
+    ///
+    /// To be used in-place of [`Hasher::finish`].
     #[inline]
+    #[must_use]
     pub fn finish<W: StableHasherResult>(self) -> W {
         W::finish(self)
     }
 
+    /// Returns the hash value for the values written so far.
+    ///
+    /// This method returns a `(u64, u64)` pair for greater precision.
+    ///
+    /// See [`StableHasher::finish`] for an idomatic use of this function.
     #[inline]
+    #[must_use]
     pub fn finalize(self) -> (u64, u64) {
         self.state.finish128()
     }
 }
 
+impl fmt::Debug for StableHasher {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        write!(f, "{:?}", self.state)
+    }
+}
+
 impl Hasher for StableHasher {
+    /// <div class="warning">
+    ///
+    /// Do not use this function, it will unconditionnaly panic.
+    ///
+    /// Use instead [`StableHasher::finish`], or [`StableHasher::finalize`]
+    /// which returns a `(u64, u64)` pair for greater precision.
```

Comment:
hm, but what if we don't need that much precision? iirc there was `Hash64` structure in rustc...

Reply:
`Hash64` just use the "lower" part in it's [implementation](https://doc.rust-lang.org/beta/nightly-rustc/src/rustc_data_structures/hashes.rs.html#61-63) of `StableHasherResult` and ignores the upper part.

I think the goal with `StableHasher::finish` is to be explicit about the use (or non use) of part of result hash and since we can't return anything else than `u64` with the default `Hasher::finish`, we have `StableHasher::{finish,finalize}` methods to do that.

In other words making `Hasher::finish` only return the lower part without the upper part is unexpected, so instead we just try to make it explicit by "forcing" the usage of `StableHasher::{finish,finalize}`.

Reply:
https://github.com/rust-lang/rustc-stable-hash/blob/cb8e141b08fb839606a5f79f9b56087cd54b764d/src/sip128.rs#L3-L5

Given this is stated as “hot”, will we gain any benefit if we had a specialized 64bit version of this stable hasher?